src/main/java/io/github/zap/commons/event/SimpleEvent.java

```diff
@@ -0,0 +1,253 @@
+
+    public SimpleEvent(@NotNull ExceptionHandler exceptionHandler, int excessCapacity) {
+        Validate.isTrue(excessCapacity >= 0, "excessCapacity cannot be negative");
+
+        this.exceptionHandler = Objects.requireNonNull(exceptionHandler, "exceptionHandler cannot be null");
+        this.excessCapacity = excessCapacity;
+        this.bakedHandlers = new EventHandler[excessCapacity];
+    }
+
+    public SimpleEvent(@NotNull ExceptionHandler exceptionHandler) {
+        this(exceptionHandler, DEFAULT_EXCESS_CAPACITY);
+    }
+
+    public SimpleEvent(@NotNull Logger logger) {
+        this(ExceptionHandlers.logHandler(logger));
+    }
+
+    public SimpleEvent() {
+        this(ExceptionHandlers.logHandler(Logger.getGlobal()));
+    }
+
+    private void addHandlerInternal(EventHandler<T> handler) {
+        if(size >= bakedHandlers.length) {
+            bakedHandlers = Arrays.copyOf(bakedHandlers, bakedHandlers.length + excessCapacity + 1);
+        }
+
+        bakedHandlers[size++] = handler;
+    }
+
+    private boolean removeHandlerInternal(EventHandler<T> handler) {
+        for(int i = 0; i < size; i++) {
+            if(bakedHandlers[i] == handler) {
+                bakedHandlers[i] = null;
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private void clearHandlersInternal() {
+        modifications.clear();
+        Arrays.fill(bakedHandlers, null);
+        size = 0;
+    }
+
+    private void rebuild() {
+        if(clearFlag) {
+            clearHandlersInternal();
+            clearFlag = false;
+        }
+
+        if(rebuildFlag) {
+            int previousNull = -1;
+            boolean hasPreviousNull = false;
+
+            int newSize = size;
+            for(int i = 0; i < size; i++) {
+                EventHandler handler = bakedHandlers[i];
+
+                if(hasPreviousNull) {
+                    if(handler != null) {
+                        bakedHandlers[previousNull] = handler;
+                        bakedHandlers[i] = null;
+                        previousNull++;
+                    }
+                    else {
+                        newSize--;
+                    }
+                }
+                else if(handler == null) {
+                    hasPreviousNull = true;
+                    previousNull = i;
+                    newSize--;
+                }
+            }
+
+            size = newSize;
+            rebuildFlag = false;
+        }
+    }
+
+    private void invokeInternal(T args) {
+        RuntimeException first = null;
+        for(int i = 0; i < size; i++) {
+            EventHandler handler = bakedHandlers[i];
+
+            try {
+                handler.invoke(this, args);
+            }
+            catch (RuntimeException exception) {
+                if(first == null) {
+                    first = exception;
+                }
+                else {
+                    first.addSuppressed(exception);
```

Comment:
Idk if `addSuppressed` method is clear here, later handlers aren't suppressing the former failed hander. I think `AggerateException` is better suited and we already have that class in `arena-api`. Here is a quote from javadocs
> Note that when one exception causes another exception, the first exception is usually caught and then the second exception is thrown in response. In other words, there is a causal connection between the two exceptions. 

The last paragraph does mention the use case of ours, but I feel like there is no semantic relationship between them. But it kinda a hacky way (take advantage) rather than a practical use case.
> Note that programmer written code is also able to take advantage of calling this method in situations where there are multiple sibling exceptions and only one can be propagated.

Reply:
i mainly like it because it's a native JDK way of handling the situation
suppressed events are formatted nicely and that is done by the JDK rather than us (which i prefer)
it's not really a 'hack' if it's explicitly mentioned in the javadoc

Reply:
You're right, but try to understand the intention of the code is a bit worse when using `AggerateException`. I will leave it up to you to choose what is the most suitable.

Reply:
i prefer addSuppressed; aggregateexception may not be as intuitive to non-C# programmers, and this way they don't have to catch a custom exception (they can catch the one that actually got thrown)

wish there was a natively supported aggregateexception though

Reply:
> aggregateexception may not be as intuitive to non-C# programmers
I usually expect a hierarchy of exceptions xd. 

> and this way they don't have to catch a custom exception
This is not true because we don't really know what exception any hander will throw.

Reply:
> I usually expect a hierarchy of exceptions xd.
addSuppressed should provide a logical order

> This is not true because we don't really know what exception any hander will throw.
we don't, but some events might be able anticipate certain kinds of exceptions (for example LoadFailureException for an event called on plugin load)